File: amazon/freq/20_Valid_Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> tmp;
        for (char c: s){   // remember
            if(c == '(' || c=='{' || c=='[') {
                tmp.push(c);
            } else if (c == ')' || c == '}' || c== ']'){
                if(tmp.empty()) return false;   //key
                if(tmp.top() != '(' && c ==')') return false;
                if(tmp.top() != '{' && c =='}') return false;
                if(tmp.top() != '[' && c ==']') return false;
                tmp.pop();
            }
        }
        return tmp.empty(); //key
    }
};

File: java/Array/20. Valid Parentheses.java
//facebook
/*
20. Valid Parentheses
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.
*/
class Solution {
    public boolean isValid(String s) {
        Stack<Character> res = new Stack();

        char[] chars = s.toCharArray();
        for (char c: chars) {
            switch (c) {
                case '}':
                    if (res.empty()) return false;
                    if (res.peek() == '{') {
                        res.pop();
                    } else {
                        return false;
                    }
                    break;
                case ']':
                    if (res.empty()) return false;
                    if (res.peek() == '[') {
                        res.pop();
                    } else {
                        return false;
                    }
                    break;
                case ')':
                    if (res.empty()) return false;
                    if (res.peek() == '(') {
                        res.pop();
                    } else {
                        return false;
                    }
                    break;
                default:
                    if (c == '(' || c == '[' || c == '{') {
                        res.push(c);
                    } else {
                        return false;
                    }
                    break;
            }
        }
        
        return res.empty();
    }
}
